.automation/build_archi_diagram.py

```diff
@@ -0,0 +1,77 @@
+# !/usr/bin/env python3
+"""
+Generate architecture diagram of Mega-Linter
+See how to do it at https://kroki.io/#how
+"""
+import base64
+import tempfile
+import zlib
+from string import Template
+
+# pylint: disable=import-error
+import requests
+
+
+class diagram:
+    GATEWAY_SERVER = "https://kroki.io"
+    DIAGRAM_TEMPLATE = "docs/assets/archi.blockdiag.txt"
+
+    def __init__(
+        self,
+        descriptor_language_count,
+        descriptor_format_count,
+        descriptor_tooling_format_count,
+    ):
+        self.diagram_dict = dict(
+            language=str(descriptor_language_count),
+            format=str(descriptor_format_count),
+            tooling_format=str(descriptor_tooling_format_count),
+        )
+
+        self.build_diagram()
+
+    # get binary image generated following format specified
+    def get_image_bin(self, format="svg", tempfile="kroki"):
+        image_content = self.run_query(
+            self.GATEWAY_SERVER, "blockdiag", format, self.encoded_diagram
+        )
+
+        output_file = tempfile + "." + format
+        with open(output_file, "wb") as file:
+            file.write(image_content)
+
+        return output_file
+
+    # get diagram textual description
+    def get_image_text(self):
+        return self.asset
+
+    # generated image on gateway (default: online kroki.io server)
+    def run_query(self, gateway, diagram_type, diagram_format, encoded_diagram):
+        request = requests.get(
+            f"{gateway}/{diagram_type}/{diagram_format}/{encoded_diagram}"
+        )
+        if request.status_code == 200:
+            return request.content
+        else:
+            raise Exception(
+                "Query failed to run by returning code of {}".format(
+                    request.status_code
+                )
+            )
+
+    # build diagram textual description and return deflate + base64 encoded version
+    def build_diagram(self):
+        with open(self.DIAGRAM_TEMPLATE, "rt") as file:
+            diagram_source = file.read().rstrip()
+
+        self.asset = Template(diagram_source).safe_substitute(self.diagram_dict)
```

Comment:
This is fairly absurd, but `safe_substitute()` is actually the more dangerous of the two methods, and [substitute()](https://docs.python.org/3/library/string.html#string.Template.substitute) is the safer of the two, because it actually raises an error if the template is malformed. Here is the relevant excerpt of the documentation of `safe_substitute()`.

> While other exceptions may still occur, this method is called “safe” because it always tries to return a usable string instead of raising an exception. In another sense, [safe_substitute()](https://docs.python.org/3/library/string.html#string.Template.safe_substitute) may be anything other than safe, since it will silently ignore malformed templates containing dangling delimiters, unmatched braces, or placeholders that are not valid Python identifiers. ~ [safe_substitute()](https://docs.python.org/3/library/string.html#string.Template.safe_substitute)

Reply:
I read the doc, and this is the reason why I've chose `safe_substitute()`. I'd like a template generated with placeholder not modified rather than an exception. Remember that it's for a diagram and error is visual for a human.
I'm ok to change it; so vote on behavior : what method should we keep ?

Reply:
I'm not seeing the drawback to using `substitute()` since the exception would make someone in the future more immediately aware if they had broken something, and hopefully show them the exact line number where they messed up. If an exception is not thrown, there is at least some risk that they don't notice they messed up (e.g., because they may not be aware what the diagram is supposed to look like).